core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigration.scala

```diff
@@ -257,13 +257,22 @@ trait WorkspaceMigrationHistory extends RawSqlQuery {
 
     final def truncate: WriteAction[Int] = sqlu"delete from #$tableName"
 
-    def startCondition(maxAttempts: Int) =
-      for {
-        count <- sql"select count(*) from #$tableName where #$startedCol is not null and #$finishedCol is null".as[Int].head
-        migrations <-
-          if (count < maxAttempts) sql"select #$allColumns from #$tableName where #$startedCol is null order by #$idCol asc limit 1".as[WorkspaceMigration]
-          else DBIO.successful(Vector.empty)
-      } yield migrations
+    // The following two definitions are `ReadWriteAction`s to make the types line up more easily
+    // in their uses
+    def getNumActiveMigrations: ReadWriteAction[Int] =
+      sql"select count(*) from #$tableName where #$startedCol is not null and #$finishedCol is null".as[Int].head
+
+    def nextMigration: ReadWriteAction[Option[(Long, UUID, Boolean)]] =
+      sql"""
+        select m.#$idCol, m.#$workspaceIdCol, w.is_locked from #$tableName m
+        join (select id, is_locked from WORKSPACE) as w on (w.id = m.#$workspaceIdCol)
+        where m.#$startedCol is null and not exists (
+            select workspace_id, status from SUBMISSION
+            where status in (#${SubmissionStatuses.activeStatuses.map("'" + _ + "'").mkString(",")})
```

Comment:
String concatenation like this in sql make me nervous. I know in this particular instance it is safe but I still don't like to have it in code for other people to copy.

So either figure out how to do this with bind parameters or leave a comment saying why it is safe to do it in this case but is generally not a good practice

Reply:
Could you explain why it's bad practice? I guess I'm not sure what to write in such a comment?
Furthermore, I'm not sure I can do this with inSetBind as we're not using that dsl for the migration table? Is that incorrect?

Reply:
Fair enough, I wasn't sure that you were referring to using raw parameters. I'll try to get it to work with bound parameters and failing that, add a comment as you suggest.